ssl/quic/quic_impl.c

```diff
@@ -4324,10 +4324,119 @@ SSL *ossl_quic_new_listener_from(SSL *ssl, uint64_t flags)
 /*
  * SSL_new_from_listener
  * ---------------------
+ * code here is derived from ossl_quic_new(). The `ssl` argument is
+ * a listener object which already come with QUIC port/engine. The newly
+ * created QUIC connection object (QCSO) is going to share the port/engine
+ * with listener (`ssl`).  The `ssl` also becomes a parent of QCSO created
+ * by this function. The caller uses QCSO instance to connect to
+ * remote QUIC server.
+ *
+ * The QCSO created here requires us to also create a channel so we
+ * can connect to remote server.
  */
 SSL *ossl_quic_new_from_listener(SSL *ssl, uint64_t flags)
 {
-    /* TODO(QUIC SERVER): Implement SSL_new_from_listener */
+    QCTX ctx;
+    QUIC_CONNECTION *qc;
+    QUIC_LISTENER *ql;
+    SSL_CONNECTION *sc = NULL;
+
+    if (flags != 0)
+        return NULL;
+
+    if (!expect_quic_listener(ssl, &ctx))
+        return NULL;
+
+    if (!SSL_up_ref(&ctx.ql->obj.ssl))
+        return NULL;
+
+    qctx_lock(&ctx);
+
+    ql = ctx.ql;
+
+    if ((qc = OPENSSL_zalloc(sizeof(*qc))) == NULL) {
+        QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_CRYPTO_LIB, NULL);
+        SSL_free(&ql->obj.ssl);
+        goto err;
+    }
+
+    /*
+     * XXX setting listener here is needed so `qc_cleanup()` does
+     * a right thing. There is `TOOD` comment in `qc_cleanup()`, which
+     * makes me believe line below is correct for now.
+     * Without this might be destroying port prematurely.
+     *
+     * Perhaps we should be grabbing a references to engine and port here.
+     */
+    qc->listener = ql;
+    qc->engine = ql->engine;
+    qc->port = ql->port;
```

Comment:
You absolutely need to grab a reference to the engine and port.

Reply:
not sure I follow here, QUIC_ENGINES and PORTS don't have reference counters, nor do other calls like ossl_quic_accept_connection. How is creating a new client SSL object via this path any different from accepting a new connection via ossl_quic_accept_connection?

That said, can we just use ossl_quic_obj_init here?

Reply:
Sorry this comment is left behind when I took this change from my working branch. Currently port, nor engine support reference counting. It is something I'd like to investigate more closely. It's perhaps just a brain-fart, given how far things got without needing a refcounter.

Reply:
> not sure I follow here, QUIC_ENGINES and PORTS don't have reference counters, nor do other calls like ossl_quic_accept_connection. How is creating a new client SSL object via this path any different from accepting a new connection via ossl_quic_accept_connection?
> 
> That said, can we just use ossl_quic_obj_init here?

the `ossl_quic_accept_conection()` is handled by `create_qc_from_incoming_conn()` snippet below comes right from there:
```
4552     if ((qc = OPENSSL_zalloc(sizeof(*qc))) == NULL) {
4553         QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_CRYPTO_LIB, NULL);
4554         SSL_free(&ql->obj.ssl);
4555         goto err;
4556     }
4557
4558     if (!ossl_quic_obj_init(&qc->obj, ql->obj.ssl.ctx,
4559                             SSL_TYPE_QUIC_CONNECTION,
4560                             &ql->obj.ssl, NULL, NULL)) {
4561         QUIC_RAISE_NON_NORMAL_ERROR(NULL, ERR_R_INTERNAL_ERROR, NULL);
4562         SSL_free(&ql->obj.ssl);
4563         goto err;
4564     }
```

In this respect the SSL_new_from_listener() is doing the same thing to set up QUIC connection object. I just added the comment so it is more obvious. Ideally we should change the `ossl_quic_obj_init()` so its behavior will become less obfuscated. May be we can have more variants like ossl_quic_obj_init() and new ossl_quic_obj_init_as_leader(). I don't know now. The comment hopes to drag attention to this, so we can change it later once we will better understand what's needed to improve it.